BlueprintUICommonControls/Sources/Internal/KeyboardObserver.swift

```diff
@@ -0,0 +1,178 @@
+ 
+ iOS Docs for keyboard management:
+ https://developer.apple.com/library/archive/documentation/StringsTextFonts/Conceptual/TextAndWebiPhoneOS/KeyboardManagement/KeyboardManagement.html
+ */
+final class KeyboardObserver {
+    
+    private let center : NotificationCenter
+    
+    weak var delegate : KeyboardObserverDelegate?
+    
+    //
+    // MARK: Initialization
+    //
+    
+    init(center : NotificationCenter = .default) {
+        
+        self.center = center
+        
+        /// We need to listen to both `will` and `keyboardDidChangeFrame` notifications. Why?
+        /// When dealing with an undocked or floating keyboard, moving the keyboard
+        /// around the screen does NOT call `willChangeFrame`; only `didChangeFrame` is called.
+        /// Before calling the delegate, we compare `old.endingFrame != new.endingFrame`,
+        /// which ensures that the delegate is notified if the frame really changes, and
+        /// prevents duplicate calls.
+
+        self.center.addObserver(self, selector: #selector(keyboardFrameChanged(_:)), name: UIWindow.keyboardWillChangeFrameNotification, object: nil)
+        self.center.addObserver(self, selector: #selector(keyboardFrameChanged(_:)), name: UIWindow.keyboardDidChangeFrameNotification, object: nil)
+    }
+    
+    private var latestNotification : NotificationInfo?
+    
+    //
+    // MARK: Handling Changes
+    //
+    
+    enum KeyboardFrame : Equatable {
+        
+        /// The current frame does not overlap the current view at all.
+        case nonOverlapping
+        
+        /// The current frame does overlap the view, by the provided rect, in the view's coordinate space.
+        case visible(frame: CGRect)
+    }
+    
+    /// How the keyboard overlaps the view provided. If the view is not on screen (eg, no window),
+    /// or the observer has not yet learned about the keyboard's position, this method returns nil.
+    func currentFrame(in view : UIView) -> KeyboardFrame? {
+        
+        guard view.window != nil else {
+            return nil
+        }
+        
+        guard let notification = self.latestNotification else {
+            return nil
+        }
+        
+        let frame = view.convert(notification.endingFrame, from: nil)
+        
+        if frame.intersects(view.bounds) {
+            return .visible(frame: frame)
+        } else {
+            return .nonOverlapping
+        }
+    }
+    
+    //
+    // MARK: Receiving Updates
+    //
+    
+    private func receivedUpdatedKeyboardInfo(_ new : NotificationInfo) {
+        
+        let old = self.latestNotification
+        
+        self.latestNotification = new
+        
+        /// Only communicate a frame change to the delegate if the frame actually changed.
+        
+        if let old = old {
+            guard old.endingFrame != new.endingFrame else {
+                return
+            }
+        }
```

Comment:
`guard` nested in `if` is a real brain-teaser. Suggest flattening to simplify:

```suggestion
        if let old = old, old.endingFrame == new.endingFrame {
            return
        }
```

Reply:
Committing locally